rust/catalyst-types/src/smt/tree.rs

```diff
@@ -0,0 +1,337 @@
+    ///
+    /// Note: This function uses the standard tree convention where height 0 is the root
+    /// and height increases toward the leaves.
+    pub fn horizontal_slice_at(
+        &self,
+        height: u8,
+    ) -> impl Iterator<Item = Result<Option<H256>, Error>> {
+        let store = self.inner.store();
+
+        horizontal_slice::<_, Hasher>(store, height)
+    }
+}
+
+impl<V> Default for Tree<V>
+where V: Default + Value + Clone
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Generates a complete horizontal slice of the Sparse Merkle Tree at a specified height.
+///
+/// This function creates an iterator representing all possible node positions at a given
+/// height, whether those nodes are materialized in storage or not.
+///
+/// # Sparse Merkle Tree Structure
+///
+/// In an SMT, each height has a theoretical maximum number of nodes:
+/// - Height 0 (root): 1 node (2^0)
+/// - Height 1: 2 nodes (2^1)
+/// - Height 2: 4 nodes (2^2)
+/// - Height h: 2^h nodes
+///
+/// However, only nodes on paths to actual leaf values are materialized. This function
+/// reconstructs the complete horizontal slice by:
+/// 1. Computing all theoretical node positions at the target height
+/// 2. Looking up which positions have materialized nodes in storage
+/// 3. For materialized nodes: computing their hash by merging left/right children
+/// 4. For non-materialized nodes: yielding `None` (representing implicit zero branches)
+///
+/// # Algorithm
+///
+/// 1. Calculate `key_prefix_length` = path length from root to target height
+/// 2. Calculate `width` = `2^key_prefix_length` (total possible nodes at this height)
+/// 3. Retrieve all materialized nodes at target height from store
+/// 4. Lazily iterate through each horizontal position (0 to width-1):
+///    - Generate the `node_key` for that position
+///    - If node is materialized: compute its hash from left/right children
+///    - If node is not materialized: yield None (virtual zero branch)
+///
+/// # Returns
+///
+/// An iterator yielding `Option<H256>` for each position in the horizontal slice:
+/// - `Some(hash)` indicates a materialized node with the computed hash
+/// - `None` indicates a virtual zero branch (non-materialized)
+///
+/// The iterator will yield `2^(ROOT_HEIGHT - target_height)` items.
+///
+/// # Performance Note
+///
+/// Returns an iterator to avoid allocating `2^key_prefix_length` elements upfront.
+/// For heights near the leaves (high values), the theoretical width can be astronomically
+/// large. The iterator allows processing nodes on-demand or early termination.
+fn horizontal_slice<V, H>(
+    store: &DefaultStore<V>,
+    target_height: u8,
+) -> impl Iterator<Item = Result<Option<H256>, Error>>
+where
+    V: Clone,
+    H: sparse_merkle_tree::traits::Hasher + Default,
+{
+    // In `sparse_merkle_tree` crate the heights are inverted (root is located at height 255).
+    #[allow(clippy::arithmetic_side_effects)]
```

Comment:
Can ignore this

Personally not prefer this arithmetic side effects. Just in case someone come and change this `ROOT_HEIGHT` or `target_height` to u32